### ISSUE TITLE
exc: The JSON element does not have the requested type. (HeapProfiler::Error)

### ISSUE THREAD

**Issue by theSoenke** [Shopify/heap-profiler#12]
Hi :wave: , when parsing a heap dump created with `ObjectSpace.dump_all` I'm getting the following error:

```
.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/parser.rb:51:in `_build_index': exc: The JSON element does not have the requested type. (HeapProfiler::Error)
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/parser.rb:51:in `build_index'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/parser.rb:63:in `build_index'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/index.rb:14:in `build!'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/index.rb:10:in `initialize'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/results.rb:79:in `new'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/results.rb:79:in `pretty_print'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/cli.rb:47:in `print_report'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/lib/heap_profiler/cli.rb:23:in `run'
	from .asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/heap-profiler-0.7.0/exe/heap-profiler:5:in `<top (required)>'
	from .asdf/installs/ruby/3.2.1/bin/heap-profiler:25:in `load'
	from .asdf/installs/ruby/3.2.1/bin/heap-profiler:25:in `<main>'

```

The heap dump is about 600MB in size and it worked fine on another project with a smaller heap dump

**Comment by casperisfine** [comment]
Hum, actually, re-reading the error, the issue isn't strictly a JSON syntax error, but that one of the values isn't of the expected type. 

To identify it you may need to bisect the heap.

**Comment by theSoenke** [comment]
@casperisfine thanks a lot. `heap-profiler clean` fixed the problem! I've been using Ruby 3.2.1 both where the heap was created and locally. In general it also seems to be a valid JSON file and parsing it line by line seemed to work.

Here's also a diff before and after the clean step:
```
1494103d1494102
< 80, "flags":{"wb_protected":true, "old":true, "uncollectible":true, "marked":true}}
1494123d1494121
< {"address":"0x7f9d07934990", "type":"STRING", "shape_id":12, "slot_size":80, "class":"0x7f9d093aed98", "frozen":true, "embedded":true, "fstring":true, "bytesize":23, "value":"indeterminate_check_box", "encoding":"UTF-8"
```

**Comment by casperisfine** [comment]
Hum, the first like is clearly only the half-end of a JSON object, I wonder what happened here.


The second line seem fine except that it's not terminated. I wonder if it might be a flushing issue of some sort.
